# DASSIM deep sequencing

Code for analysis of the DASSIM metagenomic samples.

- Analysis scripts are in `scripts/` as quarto files to generate `.html`
analysis files - but more detail in a readme in scripts folder
- stand models are in the root directory of the repo
- data is in `data_raw/` on my local machine but not pushed to github

The final analysis for the paper is generated by
`scripts/analysis_for_paper.qmd` - it needs to have had the models all fit using
the scripts below.

## Model code

All of the final models include a within-participant correlation structure to account
for repeated measurement, and are all in the root of the repo:

- `amr-gene-multilevel-gp-cholfactv2.stan` is the final logistic regression
model (AMR presence/absence)
- `negbin_gp_nopreds.stan` negative binomial absolute read count model (with offset
by sample) with no predictions generated - used for taxonomy
- `lm.stan` - linear model for Shannon Diversity

### Archived models

Not used in the final analysis

- `negbin-randintercept.stan` negative binomial absolute read count model with random
- intercept by participant
- `negbin_gp` negative binomial absolute read count model with gaussian process
- within-participant correlation 
- `poisson_gp.stan` poisson absolute read count model with gaussian process
- within-participant correlation (v bad fit)
- `amr-gene-logreg.stan` is the standard random effect logistic regression
model (AMR gene presence/absence) with random intercept by partcicipant

## Fitting the models

### AMR

- `scripts/model-amr.R` fits and saves the models - the latter scripts need these
outputs to run 
- `scripts/model-amr-diagnostics-GP.qmd` pulls and plots
- `scripts/building-correlated-random-eff-mod.qmd` outlines the AMR model and
the results from it 
- `scripts/resistome.qmd` is a description of the dataset and outputs of the models

- `scripts/model-amr_ecoli.R` fits the E. coli specific resistome model
- `scripts/model-amr-diagnostics-ECOLIGP.qmd` plots the diagnostics
- `scripts/model-amr-diagnostics-ECOLIGP.qmd` plots the diagnostics

### Taxonomy

- `scripts/fit_taxonomy_models.R` will fit the `negbin_gp_nopreds` models and save the output
- `scripts/model-taxonomy-diagnostics-GP.qmd` will use the output to plot the model
diagnostics
- `scripts/taxonomy.qmd` will plot the results


## Other scripts

`scripts/simulate_from_fitted_models.R` will run some simulations from the posterior and
generate the plots used in the FIS/ECCMID abstract,
