# DASSIM deep sequencing

Code for analysis of the DASSIM metagenomic samples.

## repo structure

- Analysis scripts are in `scripts/` as quarto files to generate `.html`
analysis files - there are a bunch of these, detailed below in "fitting
the models"
- stan models are in the root directory of the repo
- data is in `data_raw/` 

## Reproducing the paper analysis

The final analysis for the paper is generated by
`scripts/analysis_for_paper.qmd` - it needs to have had the models all fit, so
before running you will need to fit the following models, using the following
scripts

- Shannon diversity models; `scripts/fit_shannon_diversity_model.R`
- Taxonomy models: `scripts/fit_taxonomy_models.R`
- AMR models: `scripts/model-amr.R`  
- The E. coli specific AMR model: `scripts/model-amr_ecoli.R`

These will save models in `data_processed/` - the files are quite large so not
pushed to repo. Other scripts (detailed below) extract and plot diagnostics and
results as htmls.

## Fitting the models

### Dependencies

- R (tested on v4.4.2)
- Stan (tested with cmdstanR v0.8.1 using cmdstan v2.35)
- the blantyreESBL R package (v1.4.1) contains all the clinical metadata - see
[https://joelewis101.github.io/blantyreESBL/] for installation instructions.

### Model code

All of the final models include a within-participant correlation structure to account
for repeated measurement, and are all in the root of the repo:

- `amr-gene-multilevel-gp-cholfactv2.stan` is the final logistic regression
model (AMR presence/absence)
- `negbin_gp_nopreds.stan` negative binomial absolute read count model (with offset
by sample) with no predictions generated - used for taxonomy
- `lm.stan` - linear model for Shannon Diversity

### Shannon diversity

- `scripts/fit_shannon_diversity_model.R` fits and saves the models 

### AMR

- `scripts/model-amr.R` fits and saves the models - the latter scripts need these
outputs to run 
- `scripts/model-amr-diagnostics-GP.qmd` pulls and plots diagnostics
- `scripts/building-correlated-random-eff-mod.qmd` outlines the AMR model and
the results from it 
- `scripts/resistome.qmd` is a description of the dataset and outputs of the models

- `scripts/model-amr_ecoli.R` fits the E. coli specific resistome model
- `scripts/model-amr-diagnostics-ECOLIGP.qmd` plots the diagnostics
- `scripts/model-amr-diagnostics-ECOLIGP.qmd` plots the diagnostics

### Taxonomy

- `scripts/fit_taxonomy_models.R` will fit the `negbin_gp_nopreds` models and save the output
- `scripts/model-taxonomy-diagnostics-GP.qmd` will use the output to plot the model
diagnostics
- `scripts/taxonomy.qmd` will plot the results

### Archived models

Not used in the final analysis

- `negbin-randintercept.stan` negative binomial absolute read count model with random
- intercept by participant
- `negbin_gp` negative binomial absolute read count model with gaussian process
- within-participant correlation 
- `poisson_gp.stan` poisson absolute read count model with gaussian process
- within-participant correlation (v bad fit)
- `amr-gene-logreg.stan` is the standard random effect logistic regression
model (AMR gene presence/absence) with random intercept by partcicipant


## Other scripts

`scripts/simulate_from_fitted_models.R` will run some simulations from the posterior and
generate the plots used in the FIS/ECCMID abstract,
