---
title: Looking at DASSIM negative control contaminaton 
date: today
date-format: "D MMMM YYYY"
format: 
  html:
    embed-resources: true
execute:
  echo: false
  warning: false
  message: false
---

# Background

The DASSIM deep sequencing sample `NC_5_F6` and `NC_5_C6` have assembled *E.
coli* in Ed's metagenome assembly pipeline. These samples were negative
extraction controls (distilled water extracted alongside the samples, so each
`NC` sample corresponds to an extraction run. The other letters and numbers
correspond to their position in the orginal samples storage boxes:
`NC_[box]_[row][col]`. These boxes were then shipped to the UK and the final
list of samples for sequencing generated by using nanopore
concentration/contamination cutoffs to select 450 samples. The samples were
pipetted into the 96 well plates for CGR sequencing.

This document is an attempt to figure out why these samples contain quite good
*E. coli* genomes! The possibilities are:


* These negative control were contaminated with *E. coli*
* The whole extraction run was contaminated withe *E.coli* 
* There was a sample swap between negative control and another sample either in
  Malawi or in the UK while plating out.

# Negative control composition

These two negative controls have more reads than all the others, and
composition very similar to a human stool sample.

```{r} 
#| fig.cap: "Kraken composition of NC samples"

library(tidyverse)
library(phyloseq)
library(here)
library(blantyreESBL)
library(janitor)
library(readxl)
library(patchwork)
library(microViz)

biom <- import_biom(here("data_raw/vivien_taxonomy_reports/table.biom"))

samplist <-
  read_xls(here("data_raw/sample_management/submission/2021-10-15_CGR-LIMS-submission.xls")) |>
  janitor::clean_names() |>
  mutate(
    name =
      if_else(grepl("NC", sample_name),
        sample_name,
        gsub("_.+", "", sample_name)
      ),
    row = gsub("[0-9]", "", plate_position_well_a1_b1_c1_etc_or_a1_a2_a3_etc),
    col = gsub("[A-Z]", "", plate_position_well_a1_b1_c1_etc_or_a1_a2_a3_etc)
  )

# add metadata
biom <-
  ps_join(biom,
    blantyreESBL::btESBL_stoolESBL |>
      select(lab_id, pid, arm, visit, ESBL) |>
      mutate(arm = as.character(arm)) |>
      left_join(
        btESBL_participants |>
          transmute(
            pid = pid,
            abx = recieved_prehosp_ab,
            visit = 0
          ),
        by = join_by(pid, visit)
      ),
    match_sample_names = "lab_id"
  )

# mung headings
biom@tax_table@.Data <-
  substring(biom@tax_table@.Data, 4)

colnames(biom@tax_table@.Data) <-
  c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")

reads <-
  biom@tax_table@.Data |>
  as_tibble(rownames = "tax_id") |>
  left_join(
    biom@otu_table@.Data |>
      as_tibble(rownames = "tax_id"),
    by = join_by(tax_id)
  ) |>
  pivot_longer(-c(tax_id, Kingdom, Phylum, Class, Order, Family, Genus, Species)) 

reads |>
  filter(grepl("NC", name)) |>
  group_by(name, Kingdom) |>
  summarise(reads = sum(value)) |>
  ggplot(aes(name, reads, fill = fct_reorder(Kingdom, reads))) +
geom_col() +
coord_flip() +
labs(fill = "Kingdom") -> p_a

rank_bacterial_phyla <-
  reads |>
  group_by(Phylum, name) |>
  summarise(value = sum(value)) |>
  group_by(Phylum) |>
  summarise(median_relative_abundance = median(value)) |>
  arrange(-median_relative_abundance)

reads |>
  filter(grepl("NC_5_[F|C]", name)) |>
  mutate(Phylum = if_else(Phylum %in% rank_bacterial_phyla$Phylum[1:10], Phylum,
"Other")) |>
  group_by(name, Phylum) |>
  summarise(reads = sum(value)) |>
  ggplot(aes(name, reads, fill = fct_reorder(Phylum, reads))) +
geom_col() +
coord_flip() +
labs(fill = "Phylum") -> p_b

p_a / p_b

```

# Nanodrop values for negative controls

All the samples had DNA concentration quantified by nanodrop. ALl the negative
controls had concentration less than 5 ng/ul - none of the pateint samples did.
The two problematic samples also had a low concentration, but much higher number
of reads than the other negative controls. Accepting that nanodrop is rubbish, I
think that this is some evidence that there has been a sample swap.

```{r} 
#| fig.cap: "Nanodrop concentration vs number of reads - note log y scale"

reads <-
reads |>
  group_by(name) |>
  summarise(reads = sum(value)) |>
  mutate(type = if_else(grepl("NC", name), "NC", "patient sample")) |>
  left_join(
    select(
      samplist,
      c(name, plate_name_number, row, col, concentration_ng_ul)
    ),
    by = join_by(name == name)
  ) |>
  mutate(conc = as.numeric(concentration_ng_ul))

reads |>
    ggplot(aes(conc, reads, color = grepl("NC_5_[F|C]", name))) +
    geom_point() +
    facet_wrap(~ type, scales = "free") +
    labs(x = "Nanodrop conc.", color = "5_F6\nor 5_C6") +
    scale_y_log10()
    
```

# Looking at the plate

Can we see from the plate whether there has been a clear transposition of
samples? Not really!

```{r} 

reads |>
  filter(plate_name_number == 3) |>
  mutate(col = as.numeric(col)) |>
  ggplot(aes(col, row, fill = log(reads), label = name)) +
  geom_tile() +
  geom_text(size = 2) +
  scale_fill_viridis_c()

```

# Conclusions

I would guess that there has been some kind of transposition for these two
samples. This could have been when pipetting from the malawi boxes into plate
for CGR, so it may not be that the real `NC` samples were sequenced at all -
perhaps a way forward would be to remove all the low-read samples (ie below 0.5
million or some similar cutoff which would) from downstream analysis (this would
be 24 samples in total). This would likely remove the possibility that a rue
`NVC

I think we should also explore the possibility that these look at the assembled
*E. coli* from these `NC` samples and see if there are very similar *E. coli* in
the other samples on the same extraction run. A possibility would be to exclude
all the samples on the run, but I think the low nanopore concentration provides
evidence that there wasn't *lots* of contamination of these samples at the point
that they were extracted.
